### ISSUE TITLE
Can't get reproductible results using the parameters optimized from LightGBM integration

### ISSUE THREAD

**Issue by igorkf** [optuna/optuna#2886]
Hi!
After tuning the HPs using LightGBM integrations I could not create a model with similar results, using the 
optimized HPs found from Optuna.   

This is my `objective` function.   
I set a `random_seed` and `deterministic` parameters:
```python
import optuna
import optuna.integration.lightgbm as lgb


def objective(trial):
    
    # split train/test
    X_train, X_val, y_train, y_val = train_test_split(
        X,
        y,
        test_size=0.2,
        random_state=1
    )
     
    dtrain = lgb.Dataset(X_train, label=y_train)
    dval = lgb.Dataset(X_val, label=y_val)
    
    param = {
        'objective': 'multiclass',
        'num_classes': 30,
        'metric': 'multi_logloss',
        'lambda_l1': trial.suggest_loguniform('lambda_l1', 1e-8, 10.0),
        'lambda_l2': trial.suggest_loguniform('lambda_l2', 1e-8, 10.0),
        'num_leaves': trial.suggest_int('num_leaves', 2, 256),
        'feature_fraction': trial.suggest_uniform('feature_fraction', 0.4, 1.0),
        'bagging_fraction': trial.suggest_uniform('bagging_fraction', 0.4, 1.0),
        'bagging_freq': trial.suggest_int('bagging_freq', 1, 7),
        'min_child_samples': trial.suggest_int('min_child_samples', 5, 100),
        'boosting': 'gbdt',
        'random_seed': 0,
        'deterministic': True
    }    
    
    gbm = lgb.train(
        param,
        dtrain,
        valid_sets=[dtrain, dval],
        early_stopping_rounds=20
    )
    
    probs = gbm.predict(X_val)

    score = scoring_function(y_val, probs)

    return score
```

Then I run:
```
study = optuna.create_study(direction='minimize') # minimize RPS
study.optimize(objective, n_trials=1)
 
print('Number of finished trials:', len(study.trials))
print('Best trial:', study.best_trial.params)
```

After almost 3 hours (the dataset is big), optuna finished and showed me the best parameters.    

So then I trained a model using `lightgbm` with my function called `train_model`:
```python
import lightgbm as lgbm


def train_model(use_best_params=False):
    
    # split train/test
    X_train, X_val, y_train, y_val = train_test_split(
        X,
        y,
        test_size=0.2,
        random_state=1
    )
    
    dtrain = lgbm.Dataset(X_train, label=y_train)
    dval = lgbm.Dataset(X_val, label=y_val)
    
    if use_best_params:
        param = {
            'objective': 'multiclass',
            'num_classes': 30,
            'metric': 'multi_logloss',
            'lambda_l1': 0.25542763580886685,
            'lambda_l2': 3.3453145072294695e-06,
            'num_leaves': 104,
            'feature_fraction': 0.6312632028929402,
            'bagging_fraction': 0.7768384639050319,
            'bagging_freq': 3,
            'min_child_samples': 45,
            'boosting': 'gbdt',
            'random_seed': 0,
            'deterministic': True
        }
    else:
        param = {
            'objective': 'multiclass',
            'num_classes': 30,
            'metric': 'multi_logloss',
            'boosting': 'gbdt',
            'random_seed': 0,
            'deterministic': True
        } 
    
    evals_result = {}
    clf = lgbm.train(
        param,
        dtrain,
        valid_sets=[dtrain, dval],
        early_stopping_rounds=20,
        evals_result=evals_result
    )
    
    # predicting
    probs_train = clf.predict(X_train)
    probs_val = clf.predict(X_val)
    
    score_train = scoring_function(y_train, probs_train)
    score_val = scoring_function(y_val, probs_val)
    print('SCORE(train):', score_train)
    print('SCORE (val):', score_val)

    return clf, evals_result
```

Then I run:
```
clf_best, evals_result_best = train_model(use_best_params=True)
```

And for my surprise, the model is VERY different from the model found with optuna.     
The `multi_logloss` are very very different, and this model is overfitting too much.   
The model found from optuna was good for me (the train and validation loss were good) and I would like
to use this model to make predictions to new data.    

Am I'm doing something wrong?   
Sorry for the long answer, but I'm trying to figure out this for a long time now.   
Thanks!

**Comment by nzw0301** [comment]
Unfortunately, the code was not reproducible for me; several undefined functions. Moreover, I think this issue is not Optuna's issue I'm afraid.

**Comment by nzw0301** [comment]
When I ran the following code by modifying your code, the learnt model's validation accuracy was the same as the optuna's result. Thus please consider double-checking the definition of evaluation metric, the logic of comparison of the evaluation scores. I'd suggest using a much smaller training dataset for debugging.

```python
import optuna
import lightgbm as lgb
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score



def objective(trial):

    # split train/test
    X, y = load_iris(return_X_y=True)
    X_train, X_val, y_train, y_val = train_test_split(
        X,
        y,
        test_size=0.2,
        random_state=1
    )

    dtrain = lgb.Dataset(X_train, label=y_train)
    dval = lgb.Dataset(X_val, label=y_val)

    param = {
        'objective': 'multiclass',
        'num_classes': len(np.unique(y)),
        'metric': 'multi_logloss',
        'lambda_l1': trial.suggest_loguniform('lambda_l1', 1e-8, 10.0),
        'lambda_l2': trial.suggest_loguniform('lambda_l2', 1e-8, 10.0),
        'num_leaves': trial.suggest_int('num_leaves', 2, 256),
        'feature_fraction': trial.suggest_uniform('feature_fraction', 0.4, 1.0),
        'bagging_fraction': trial.suggest_uniform('bagging_fraction', 0.4, 1.0),
        'bagging_freq': trial.suggest_int('bagging_freq', 1, 7),
        'min_child_samples': trial.suggest_int('min_child_samples', 1, 100),
        'boosting': 'gbdt',
        'random_seed': 0,
        'deterministic': True
    }

    gbm = lgb.train(
        param,
        dtrain,
        valid_sets=[dtrain, dval],
        early_stopping_rounds=20
    )

    probs = gbm.predict(X_val)
    pred_labels = np.argmax(probs, axis=1)
    score = accuracy_score(y_val, pred_labels)

    return score


study = optuna.create_study(direction='maximize', sampler=optuna.samplers.TPESampler(seed=7))
study.optimize(objective, n_trials=20)


def train_model():
    X, y = load_iris(return_X_y=True)
    # split train/test
    X_train, X_val, y_train, y_val = train_test_split(
        X,
        y,
        test_size=0.2,
        random_state=1
    )

    dtrain = lgb.Dataset(X_train, label=y_train)
    dval = lgb.Dataset(X_val, label=y_val)

    param = {
        'objective': 'multiclass',
        'num_classes': 3,
        'metric': 'multi_logloss',
        # this params from Optuna's optimisation
        'lambda_l1': 4.861548294339897e-08, 'lambda_l2': 0.10453678278574731, 'num_leaves': 113, 'feature_fraction': 0.8340791066985647, 'bagging_fraction': 0.9867937071979616, 'bagging_freq': 4, 'min_child_samples': 51,
        'boosting': 'gbdt',
        'random_seed': 0,
        'deterministic': True
    }

    evals_result = {}
    clf = lgb.train(
        param,
        dtrain,
        valid_sets=[dtrain, dval],
        early_stopping_rounds=20,
        evals_result=evals_result
    )

    probs = clf.predict(X_val)
    pred_labels = np.argmax(probs, axis=1)
    score = accuracy_score(y_val, pred_labels)

    return score

assert study.best_value == train_model()
```

**Comment by nzw0301** [comment]
By the way, I realised the first code block can be optimised because `optuna.integration.lightgbm.train` is wrapped function that search hyper-parameter inside this function. So we need neither objectives function that samples hyper-parameters nor creation of `study`:


```python
import optuna.integration.lightgbm as lgb


X_train, X_val, y_train, y_val = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=1
)

dtrain = lgb.Dataset(X_train, label=y_train)
dval = lgb.Dataset(X_val, label=y_val)

param = {
    'objective': 'multiclass',
    'num_classes': 3,
    'metric': 'multi_logloss',
    'boosting': 'gbdt',
    'random_seed': 0,
    'deterministic': True
}

gbm = lgb.train(
    param,
    dtrain,
    valid_sets=[dtrain, dval],
    early_stopping_rounds=20
)

# evaluation
...
```

**Comment by igorkf** [comment]
Hi, @nzw0301 
Thank you very much for the response.
You are right...I messed up the things a little bit!

I was following this [tutorial](https://medium.com/optuna/lightgbm-tuner-new-optuna-integration-for-hyperparameter-optimization-8b7095e99258), but I was getting the code from another source and then I confused myself.

The problem is solved. Thanks.

**Comment by beifanglingyang** [comment]
> Hi, @nzw0301 Thank you very much for the response. You are right...I messed up the things a little bit!
> 
> I was following this [tutorial](https://medium.com/optuna/lightgbm-tuner-new-optuna-integration-for-hyperparameter-optimization-8b7095e99258), but I was getting the code from another source and then I confused myself.
> 
> The problem is solved. Thanks.


Hello, igorf, I am a beginner who is new to optuna. Could you please tell me how to solve the problem you encountered?  I encountered a similar problem but I didn't know how to solve it. Thanks very much!

**Comment by igorkf** [comment]
> > Hi, @nzw0301 Thank you very much for the response. You are right...I messed up the things a little bit!
> > I was following this [tutorial](https://medium.com/optuna/lightgbm-tuner-new-optuna-integration-for-hyperparameter-optimization-8b7095e99258), but I was getting the code from another source and then I confused myself.
> > The problem is solved. Thanks.
> 
> Hello, igorf, I am a beginner who is new to optuna. Could you please tell me how to solve the problem you encountered? I encountered a similar problem but I didn't know how to solve it. Thanks very much!

hi @beifanglingyang , I'm not sure about your problem, but you can try follow the tutorial I was talking about.
